File: httpserver/src/main/java/com/github/mjdev/libaums/server/http/server/AsyncHttpServer.java
package com.github.mjdev.libaums.server.http.server;

import android.util.Log;

import com.github.mjdev.libaums.fs.UsbFile;
import com.github.mjdev.libaums.fs.UsbFileInputStream;
import com.github.mjdev.libaums.server.http.UsbFileProvider;
import com.github.mjdev.libaums.server.http.exception.NotAFileException;
import com.koushikdutta.async.AsyncServer;
import com.koushikdutta.async.http.server.AsyncHttpServerRequest;
import com.koushikdutta.async.http.server.AsyncHttpServerResponse;
import com.koushikdutta.async.http.server.HttpServerRequestCallback;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;


/**
 * Created by magnusja on 16/12/16.
 */
public class AsyncHttpServer implements HttpServer, HttpServerRequestCallback {
    private static final String TAG = AsyncHttpServer.class.getSimpleName();

    private UsbFileProvider usbFileProvider;
    private com.koushikdutta.async.http.server.AsyncHttpServer server =
            new com.koushikdutta.async.http.server.AsyncHttpServer();
    private boolean isAlive = false;
    private int port;

    public AsyncHttpServer(int port) {
        this.port = port;

        server.get("/.*", this);
    }

    @Override
    public void start() throws IOException {
        server.listen(port);
        isAlive = true;
    }

    @Override
    public void stop() throws IOException {
        server.stop();
        // force the server to stop even if there are ongoing connections
        AsyncServer.getDefault().stop();
        isAlive = false;
    }

    @Override
    public boolean isAlive() {
        return isAlive;
    }

    @Override
    public String getHostname() {
        return null;
    }

    @Override
    public int getListeningPort() {
        return port;
    }

    @Override
    public void setUsbFileProvider(UsbFileProvider provider) {
        usbFileProvider = provider;
    }

    @Override
    public void onRequest(AsyncHttpServerRequest request, AsyncHttpServerResponse response) {
        String uri;
        try {
            uri = URLDecoder.decode(request.getPath(), "utf-8");
        } catch (UnsupportedEncodingException e) {
            Log.e(TAG, "could not decode URL", e);
            response.code(404);
            response.send(e.getMessage());
            return;
        }

        Log.d(TAG, "Uri: " + uri);

        try {
            UsbFile fileToServe = usbFileProvider.determineFileToServe(uri);
            response.sendStream(new UsbFileInputStream(fileToServe), fileToServe.getLength());
        } catch (FileNotFoundException e) {
            response.code(404);
            response.send(e.getMessage());
        } catch (NotAFileException e) {
            response.code(400);
            response.send(e.getMessage());
        } catch (IOException e) {
            response.code(500);
            response.send(e.getMessage());
        }
    }
}

File: httpserver/src/main/java/com/github/mjdev/libaums/server/http/server/AsyncHttpServer.kt
package com.github.mjdev.libaums.server.http.server

import android.util.Log
import com.github.mjdev.libaums.fs.UsbFileInputStream
import com.github.mjdev.libaums.server.http.UsbFileProvider
import com.github.mjdev.libaums.server.http.exception.NotAFileException
import com.koushikdutta.async.AsyncServer
import com.koushikdutta.async.http.server.AsyncHttpServerRequest
import com.koushikdutta.async.http.server.AsyncHttpServerResponse
import com.koushikdutta.async.http.server.HttpServerRequestCallback
import java.io.FileNotFoundException
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URLDecoder


/**
 * Created by magnusja on 16/12/16.
 */
class AsyncHttpServer(private val port: Int) : HttpServer, HttpServerRequestCallback {

    override lateinit var usbFileProvider: UsbFileProvider
    private val server = com.koushikdutta.async.http.server.AsyncHttpServer()
    override var isAlive = false
    override val hostname: String
        get() = ""
    override val listeningPort: Int
        get() = port

    init {

        server.get("/.*", this)
    }

    @Throws(IOException::class)
    override fun start() {
        server.listen(port)
        isAlive = true
    }

    @Throws(IOException::class)
    override fun stop() {
        server.stop()
        // force the server to stop even if there are ongoing connections
        AsyncServer.getDefault().stop()
        isAlive = false
    }

    override fun onRequest(request: AsyncHttpServerRequest, response: AsyncHttpServerResponse) {
        val uri: String
        try {
            uri = URLDecoder.decode(request.path, "utf-8")
        } catch (e: UnsupportedEncodingException) {
            Log.e(TAG, "could not decode URL", e)
            response.code(404)
            response.send(e.message)
            return
        }

        Log.d(TAG, "Uri: $uri")

        try {
            val fileToServe = usbFileProvider.determineFileToServe(uri)
            response.sendStream(UsbFileInputStream(fileToServe), fileToServe.length)
        } catch (e: FileNotFoundException) {
            response.code(404)
            response.send(e.message)
        } catch (e: NotAFileException) {
            response.code(400)
            response.send(e.message)
        } catch (e: IOException) {
            response.code(500)
            response.send(e.message)
        }

    }

    companion object {
        private val TAG = AsyncHttpServer::class.java.simpleName
    }
}
